### ISSUE TITLE
Interested in this refactor?

### ISSUE THREAD

**Issue by atrauzzi** [thephpleague/oauth2-server#314]
Hey @alexbilbie -- As promised, I've been working on a refactor.  Having looked at both yours and bshaffer's Oauth plugins, I figured there's an opportunity to make something a little more user friendly.
I'd love your input before I go too much further and maybe to see if you're interested in adopting these changes as a new foundation for oauth2-server.

Before getting into things, here are some notes to keep in mind..

 - It's not 100% PSR-2, but I've no probelms making it such for a merge.
 - I don't oppose testing, but I was more interested in just getting the refactor done.  So they're not in here, but of course tests can be remade or retooled for the new structure.
 - All examples have been removed as the touchpoints have changed.
 - Configuration is centralized to one value object that sports fluent setters.
 - I'm still debating whether there should be domain services.  There's already one for Scopes, if I accidentally left anything behind on the cutting room floor, it probably belongs in one. :)
 - Entities are now strictly interfaces with some optional traits available for non-storage-specific entity business logic.
 - I plan on making a comprehensive Laravel integration after I know where this refactor is going to live.

https://github.com/atrauzzi/oauth2-server/tree/cleanup

Regardless of what you decide, I'd really like to thank you for your work so far on this library.  Although I do hope you like what I've done enough to use it as a new starting point.

**Comment by alexbilbie** [comment]
Please bear with me, I'm exceedingly busy for at least the next few days.




I will get to this.



—

**Comment by atrauzzi** [comment]
Naturally!

I'll take that time to double check things in the library.  I just wanted to let you know more formally that I'm ready to collaborate!

**Comment by atrauzzi** [comment]
Just an update for tonight, I've pushed a slew of improvements to the entities and repositories that make them much more sensible for use with non-ORM back-ends.  This is all as I continue work on my Laravel package based on these changes which will use the cache drivers to store things.

**Comment by alexbilbie** [comment]
I've got a bit of time so I'm going to take a look at this

**Comment by atrauzzi** [comment]
If you are interested in seeing an implementation of the interfaces, you can watch here: https://github.com/atrauzzi/laravel-oauth2-server

I think it'll mostly be obvious to you, but just in case it helps.  All the repositories and entities are made.

(I'm currently getting started on the dreaded authorization code redirect flows.  Looking to get it all tied into Laravel's auth interfaces where applicable.)

**Comment by alexbilbie** [comment]
I've done my best to read through the code and here is my feedback. Please don't take any of this personally, they're just my gut reactions.

1. New terms such as "flows", "contracts" and "oauthable" are confusing and unnecessary.
2. A grant is not a contract, it is a procedure.
3. The global config object doesn't sit right because you want to tweak properties of different entities (e.g. setting different access token TTLs for different grant types).
4. Most of the changes seem to be renaming of methods and properties and just moving things around unnecessarily.

The problem with this project is that it got a bit too "heavy", it can be time consuming to implement all of the "storage interfaces" (which I agree should actually be called repository interfaces) and the actual execution path of going from `CLIENT: here are my credentials` to `AUTH SERVER: here's an access token` is too complex.

With the lack of readability of your code (whats with all the extra tab spacing?), no examples or explanations as to why things have changed I don't see anything at this moment that I want to pull in.

In my opinion the right direction is to go down a command and event driven model with lightweight, easy to implement repository interfaces.

**Comment by atrauzzi** [comment]
Not a problem, part of me suspected this is what would happen and I do definitely disagree with your appraisal based on the current complexity and rigidity of the library.

I'm going to address your points in order to keep things clear in this ticket for future readers:

1. Oauthable is the interface anything you want to authorize by implements.  This isn't unusual in any way.  This is just how you get rid of "sessions" which *is* unusual, specifically because it does not pertain to anything in the oauth spec.  If the term bothers you, check out Authenticatable in Laravel: https://github.com/laravel/laravel/blob/master/app/User.php
The term "flow" appers in normal Oauth terminology, nothing unusual there, so I'm not sure what you're basing that assumption off of.

2. You will also see the use of the term "contract" in the link above it's just a normal-language non-keyword term.  Not confusing, just programmers communicating better.

3. Easily remedied by adding a configuration for each combination.  Certainly more desirable than having all that disparate state floating around in various classes.  You create more busywork for the implementer by having so many touchpoints.

4. That's exactly why I selected your package as a baseline.  The one real gem was how you assembled flows and tokens.
What I did here was make it friendly to different types of storage layers and less rigid by loosening a lot of coupling.  I'd make a friendly wager that changing to commands and events will only add complexity with the way things are being approached right now.


Also, I apologize, some of the indentation is a bit messed up because I didn't hit all the files with a space-to-tab conversion - that's because I use tabs over spaces and anticipate people to use an editor that allows them to configure tab length.  I'll get the few stragglers once I spin up a new repository for the code and re-commit everything.

If you get proper look in at any point and change your mind, do get in touch.  I hate fragmentation just as much as the next guy.
